openshift-kube-apiserver/admission/customresourcevalidation/node/restrict_extreme_worker_latency_profile.go

```diff
@@ -0,0 +1,122 @@
+package node
+
+import (
+	"fmt"
+	"io"
+
+	"k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/admission"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"k8s.io/kubernetes/openshift-kube-apiserver/admission/customresourcevalidation"
+)
+
+var rejectionScenarios = []struct {
+	fromProfile configv1.WorkerLatencyProfileType
+	toProfile   configv1.WorkerLatencyProfileType
+}{
+	{fromProfile: "", toProfile: configv1.LowUpdateSlowReaction},
+	{fromProfile: configv1.LowUpdateSlowReaction, toProfile: ""},
+	{fromProfile: configv1.DefaultUpdateDefaultReaction, toProfile: configv1.LowUpdateSlowReaction},
+	{fromProfile: configv1.LowUpdateSlowReaction, toProfile: configv1.DefaultUpdateDefaultReaction},
+}
+
+const PluginName = "config.openshift.io/RestrictExtremeWorkerLatencyProfile"
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return customresourcevalidation.NewValidator(
+			map[schema.GroupResource]bool{
+				configv1.Resource("nodes"): true,
+			},
+			map[schema.GroupVersionKind]customresourcevalidation.ObjectValidator{
+				configv1.GroupVersion.WithKind("Node"): configNodeV1{},
+			})
+	})
+}
+
+func toConfigNodeV1(uncastObj runtime.Object) (*configv1.Node, field.ErrorList) {
+	if uncastObj == nil {
+		return nil, nil
+	}
+
+	allErrs := field.ErrorList{}
+
+	obj, ok := uncastObj.(*configv1.Node)
+	if !ok {
+		return nil, append(allErrs,
+			field.NotSupported(field.NewPath("kind"), fmt.Sprintf("%T", uncastObj), []string{"Node"}),
+			field.NotSupported(field.NewPath("apiVersion"), fmt.Sprintf("%T", uncastObj), []string{"config.openshift.io/v1"}))
+	}
+
+	return obj, nil
+}
+
+type configNodeV1 struct{}
+
+func validateConfigNodeForExtremeLatencyProfile(obj, oldObj *configv1.Node) *field.Error {
+	fromProfile := oldObj.Spec.WorkerLatencyProfile
+	toProfile := obj.Spec.WorkerLatencyProfile
+
+	for _, rejectionScenario := range rejectionScenarios {
+		if fromProfile == rejectionScenario.fromProfile && toProfile == rejectionScenario.toProfile {
+			return field.Invalid(field.NewPath("spec", "workerLatencyProfile"), obj.Spec.WorkerLatencyProfile,
+				fmt.Sprintf(
+					"cannot update worker latency profile from %q to %q as extreme profile transition is unsupported",
```

Comment:
Shouldn't we suggest that they should choose other profile with less extreme transition and wait for it to complete before choosing another transition?

The user still can choose Middle and Low in quick succession to circumvent this validation, but that might be okay if we give them  some more descriptive warning as this one. Plus we have also the validation in the operators to enforce this and not damage the cluster.

Reply:
As per our discussion, we can suggest the user to use a less extreme profile and since the validation check for Default -> Medium -> Low in a very short while is difficult to keep track of operator progress from within admission, we are skipping that as already our operator logic has the latency profile config suppressor(s) to deal with such situations. [1] [2]

[1] https://github.com/openshift/library-go/pull/1364
[2] https://github.com/openshift/library-go/pull/1372

Reply:
@swghosh looks good, but the string concatenation is  unnecessary ( `+` )